src/crimson/osd/shard_services.cc

```diff
@@ -93,14 +94,73 @@ seastar::future<> PerShardState::broadcast_map_to_pgs(
   return seastar::parallel_for_each(
     pgs.begin(), pgs.end(),
     [=, &shard_services](auto& pg) {
+      auto old_map = pg.second->get_osdmap();
       return shard_services.start_operation<PGAdvanceMap>(
 	pg.second,
 	shard_services,
 	epoch,
-	PeeringCtx{}, false).second;
+	PeeringCtx{}, false, false).second.then(
+     [this, old_map, epoch, &shard_services, &pg] {
+      return identify_splits(
+	shard_services,
+	pg.second,
+	old_map,
+	epoch).then(
+	  [this, old_map, epoch, &shard_services, &pg] (std::tuple<std::set<spg_t>, bool> &&ret) {
+	  auto children = std::get<0>(ret);
+	  bool is_split = std::get<1>(ret);
+	  if (is_split) {
+	    return shard_services.get_map(epoch).then(
+	      [this, &shard_services, &pg, children] (cached_map_t&& new_map) {
+		return shard_services.start_operation<PGSplitting>(
+		  pg.second,
+		  shard_services,
+		  new_map,
+		  std::move(children),
+		  PeeringCtx{}).second;
+	    });
+	  } else {
+	    return seastar::now();
+	  }
+        });
+      return seastar::now();
     });
+    return seastar::now();
+  });
+}
+
+seastar::future<std::tuple<std::set<spg_t>, bool>> PerShardState::identify_splits(
+  ShardServices &shard_services,
+  Ref<PG> pg,
+  cached_map_t cur_map,
+  epoch_t epoch)
+{
+  LOG_PREFIX(PerShardState::identify_splits);
+  unsigned old_pg_num;
+  if (cur_map->have_pg_pool(pg->get_pgid().pool())) {
+    old_pg_num = cur_map->get_pg_num(pg->get_pgid().pool());
```

Comment:
I'm not sure that comparing `cur_map` to `new_map` is enough. I think that we must use `pg_num_history` here to keep the history in place in terms of splitting/merging and so on.